narwhals/utils.py

```diff
@@ -297,14 +302,40 @@ def maybe_set_index(df: T, column_names: str | list[str]) -> T:
         4  1
         5  2
     """
+
     df_any = cast(Any, df)
     native_frame = to_native(df_any)
+
+    if column_names is not None and index is not None:
+        msg = "Only one of `column_names` or `keys` should be provided"
+        raise ValueError(msg)
+
+    if not column_names and not index:
+        msg = "Either `column_names` or `keys` should be provided"
+        raise ValueError(msg)
+
     if is_pandas_like_dataframe(native_frame):
-        return df_any._from_compliant_dataframe(  # type: ignore[no-any-return]
-            df_any._compliant_frame._from_native_frame(
-                native_frame.set_index(column_names)
+        if column_names is not None:
+            return df_any._from_compliant_dataframe(  # type: ignore[no-any-return]
+                df_any._compliant_frame._from_native_frame(
+                    native_frame.set_index(column_names)
+                )
             )
-        )
+
+        if index is not None:
+            from narwhals.series import Series
+
+            if _is_iterable(index):
+                index = [
+                    key.to_native() if isinstance(key, Series) else key for key in index
+                ]
+            if isinstance(index, Series):
+                index = index.to_native()
```

Comment:
How about the following? (will require formatting and maybe ignoring mypy)

```suggestion
        index = [to_native(key, pass_through=True) for key in index] if _is_iterable(index) else [to_native(index, pass_through=True)]
```